### DISCUSSION TITLE
Two users using taipy

### DISCUSSION THREAD

**Discussion by vfreysz** [Avaiga/taipy#101] [Announcements]
Hello,

What happens if two users use the service at the same time ?

If they edit a value on two different computer ?

How to handle this case ?

**Comment by FlorianJacta** [comment]
**Taipy GUI**
In Taipy GUI, each session/user has its own State ([doc](https://docs.taipy.io/en/latest/manuals/reference/taipy.gui.State/)). In this State, all the variables present on the GUI are accessible. Changing a data in a State will not change this variable in another State of another session. The [Getting Started](https://docs.taipy.io/en/release-2.1/getting_started/getting-started-gui/step_03/ReadMe/#multi-user-state) for GUI talks about this in more detail.

**Taipy Core**

In Taipy Core, the concurrency is handled by Taipy directly through the Data Nodes and the Scheduler. When the first user writes a data node, Taipy locks the data node until the writing is done. The writing is rejected if a second user tries to write the same data node while it is locked. Note that the locked status of the data node is exposed to the users so they can know if it is locked.
